body/index.ts

```diff
@@ -0,0 +1,112 @@
+export type ReadBodyOpts = {
+  limit: number;
+  expectedLength?: number;
+};
+
+type EventHandlerLike = (
+  event: string,
+  listener: (...args: any[]) => void,
+) => void;
+
+// The fields from stream.Readable that we use
+export interface ReadableStreamLike {
+  on?: EventHandlerLike;
+  removeListener?: EventHandlerLike;
+  readable?: boolean;
+}
+
+export async function readBody(
+  stream: ReadableStreamLike,
+  opts: ReadBodyOpts,
+): Promise<string> {
+  const decoder = new TextDecoder("utf-8");
+  let buffer = "";
+  let complete = false;
+  let received = 0;
+  const limit = opts.limit;
+  if (typeof limit !== "number") {
+    return Promise.reject(new Error("must set a limit"));
+  }
+  const length = opts.expectedLength || null;
+
+  if (typeof stream.readable !== "undefined" && !stream.readable) {
+    return Promise.reject(new Error("stream is not readable"));
+  }
+  if (typeof stream.on !== "function") {
+    return Promise.reject(new Error("missing `on` function"));
+  }
+  if (typeof stream.removeListener !== "function") {
+    return Promise.reject(new Error("missing `removeListener` function"));
+  }
+
+  return new Promise((resolve, reject) => {
+    // This was already checked at the top of the function but TypeScript lost
+    // the context
+    if (typeof stream.on === "function") {
+      stream.on("aborted", onAborted);
+      stream.on("close", cleanup);
+      stream.on("data", onData);
+      stream.on("end", onEnd);
+      stream.on("error", onEnd);
+    }
+
+    function done(err?: Error, buffer?: string) {
+      // Ensure we avoid double resolve/reject if called more than once
+      if (complete) return;
+
+      complete = true;
+
+      cleanup();
+      if (typeof err !== "undefined") {
+        reject(err);
+      } else if (typeof buffer !== "undefined") {
+        resolve(buffer);
+      }
+    }
+
+    function onAborted() {
+      done(new Error("Stream was aborted"));
+    }
+
+    function onData(chunk: Buffer) {
+      received += chunk.length;
+
+      if (received > limit) {
+        done(new Error("request entity too large"));
+      } else {
+        buffer += decoder.decode(chunk, { stream: true });
```

Comment:
The hover docs state that:
> Returns the result of running encoding's decoder. The method can be invoked zero or more times with options's stream set to true, and then once without options's stream (or set to false), to process a fragmented input. If the invocation without options's stream (or set to false) has no input, it's clearest to omit both arguments.
>
```
var string = "", decoder = new TextDecoder(encoding), buffer;
while(buffer = next_chunk()) {
  string += decoder.decode(buffer, {stream:true});
}
string += decoder.decode(); // end-of-queue
```
> If the error mode is "fatal" and encoding's decoder returns error, throws a TypeError.

It seems that you need to call `decoder.decode()` once upon completion otherwise you might not get everything that is buffered or an error.